From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Christian=20K=C3=B6nig?= <christian.koenig@amd.com>
Date: Tue, 23 Nov 2021 09:33:07 +0100
Subject: dma-buf & drm/amdgpu: remove dma_resv workaround

Rework the internals of the dma_resv object to allow adding more than one
write fence and remember for each fence what purpose it had.

This allows removing the workaround from amdgpu which used a container for
this instead.

Signed-off-by: Christian König <christian.koenig@amd.com>
Reviewed-by: Daniel Vetter <daniel.vetter@ffwll.ch>
Cc: amd-gfx@lists.freedesktop.org
Link: https://patchwork.freedesktop.org/patch/msgid/20220407085946.744568-4-christian.koenig@amd.com

(Backported from commit 047a1b877ed48098bed71fcfb1d4891e1b54441d linux)

Reviewed-by: Arvind Yadav <arvind.yadav@intel.com>
Signed-off-by: Varun Gupta <varun.gupta@intel.com>
---
 drivers/dma-buf/dma-resv.c | 65 +++++++++++++++++++++++++++++++-------
 include/linux/dma-resv.h   |  3 ++
 2 files changed, 57 insertions(+), 11 deletions(-)

diff --git a/drivers/dma-buf/dma-resv.c b/drivers/dma-buf/dma-resv.c
index bcce96d4a753..67da04653311 100644
--- a/drivers/dma-buf/dma-resv.c
+++ b/drivers/dma-buf/dma-resv.c
@@ -74,6 +74,43 @@ static struct dma_resv_list *dma_resv_list_alloc(unsigned int shared_max)
 	return list;
 }
 
+/**
+ * dma_resv_list_entry - extract fence and usage from a shared[] slot
+ * @list: the shared fence list
+ * @index: which slot to read
+ * @resv: the dma_resv object, or NULL when only RCU is held
+ * @fence: where to store the untagged fence pointer
+ * @usage: where to store the usage, or NULL if not needed
+ */
+static void dma_resv_list_entry(struct dma_resv_list *list, unsigned int index,
+				struct dma_resv *resv, struct dma_fence **fence,
+				enum dma_resv_usage *usage)
+{
+	long tmp;
+
+	tmp = (long)rcu_dereference_check(list->shared[index],
+					  resv ? dma_resv_held(resv) : true);
+	*fence = (struct dma_fence *)(tmp & ~DMA_RESV_LIST_MASK);
+	if (usage)
+		*usage = tmp & DMA_RESV_LIST_MASK;
+}
+
+/**
+ * dma_resv_list_set - write a fence and its usage into a shared[] slot
+ * @list: the shared fence list
+ * @index: which slot to write
+ * @fence: the fence to store
+ * @usage: the usage to encode into the low two bits
+ */
+static void dma_resv_list_set(struct dma_resv_list *list, unsigned int index,
+			      struct dma_fence *fence,
+			      enum dma_resv_usage usage)
+{
+	long tmp = ((long)fence) | usage;
+
+	RCU_INIT_POINTER(list->shared[index], (struct dma_fence *)tmp);
+}
+
 /**
  * dma_resv_list_free - free fence list
  * @list: list to free
@@ -87,8 +124,12 @@ static void dma_resv_list_free(struct dma_resv_list *list)
 	if (!list)
 		return;
 
-	for (i = 0; i < list->shared_count; ++i)
-		dma_fence_put(rcu_dereference_protected(list->shared[i], true));
+	for (i = 0; i < list->shared_count; ++i) {
+		struct dma_fence *fence;
+
+		dma_resv_list_entry(list, i, NULL, &fence, NULL);
+		dma_fence_put(fence);
+	}
 
 	kfree_rcu(list, rcu);
 }
@@ -169,14 +210,14 @@ int dma_resv_reserve_shared(struct dma_resv *obj, unsigned int num_fences)
 	 * the new.
 	 */
 	for (i = 0, j = 0, k = max; i < (old ? old->shared_count : 0); ++i) {
+		enum dma_resv_usage usage;
 		struct dma_fence *fence;
 
-		fence = rcu_dereference_protected(old->shared[i],
-						  dma_resv_held(obj));
+		dma_resv_list_entry(old, i, obj, &fence, &usage);
 		if (dma_fence_is_signaled(fence))
-			RCU_INIT_POINTER(new->shared[--k], fence);
+			dma_resv_list_set(new, --k, fence, usage);
 		else
-			RCU_INIT_POINTER(new->shared[j++], fence);
+			dma_resv_list_set(new, j++, fence, usage);
 	}
 	new->shared_count = j;
 
@@ -197,8 +238,7 @@ int dma_resv_reserve_shared(struct dma_resv *obj, unsigned int num_fences)
 	for (i = k; i < max; ++i) {
 		struct dma_fence *fence;
 
-		fence = rcu_dereference_protected(new->shared[i],
-						  dma_resv_held(obj));
+		dma_resv_list_entry(new, i, obj, &fence, NULL);
 		dma_fence_put(fence);
 	}
 	kfree_rcu(old, rcu);
@@ -306,9 +346,12 @@ void dma_resv_add_excl_fence(struct dma_resv *obj, struct dma_fence *fence)
 	write_seqcount_end(&obj->seq);
 
 	/* inplace update, no shared fences */
-	while (i--)
-		dma_fence_put(rcu_dereference_protected(old->shared[i],
-						dma_resv_held(obj)));
+	while (i--) {
+		struct dma_fence *old_shared;
+
+		dma_resv_list_entry(old, i, obj, &old_shared, NULL);
+		dma_fence_put(old_shared);
+	}
 
 	dma_fence_put(old_fence);
 }
diff --git a/include/linux/dma-resv.h b/include/linux/dma-resv.h
index 3520045a280c..4edff7e44f82 100644
--- a/include/linux/dma-resv.h
+++ b/include/linux/dma-resv.h
@@ -60,6 +60,9 @@ struct dma_resv_list {
 	struct dma_fence __rcu *shared[];
 };
 
+/* Mask for the lower fence pointer bits */
+#define DMA_RESV_LIST_MASK	0x3
+
 /**
  * struct dma_resv - a reservation object manages fences for a buffer
  * @lock: update side lock
-- 
2.43.0

